File: src/main/java/dmgcalculator/DmgCalculatorMod.java
package dmgcalculator;

import com.badlogic.gdx.Files;
import com.badlogic.gdx.backends.lwjgl.LwjglFileHandle;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.evacipated.cardcrawl.modthespire.Loader;
import com.evacipated.cardcrawl.modthespire.ModInfo;
import com.evacipated.cardcrawl.modthespire.Patcher;
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer;
import com.megacrit.cardcrawl.cards.AbstractCard;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.scannotation.AnnotationDB;

import java.util.Arrays;
import java.util.Collections;
import java.util.Optional;
import java.util.Set;

import basemod.BaseMod;
import basemod.interfaces.PostInitializeSubscriber;
import basemod.interfaces.PostRenderSubscriber;
import dmgcalculator.interfaces.OnCardHoveredSubscriber;
import dmgcalculator.publisher.CardHoverPublisher;
import dmgcalculator.renderer.CalculateIncomingDamageRenderer;
import dmgcalculator.renderer.CalculateOutgoingDamageRenderer;
import dmgcalculator.util.GeneralUtils;
import dmgcalculator.util.TextureLoader;
import dmgcalculator.util.Utils;

@SpireInitializer
public class DmgCalculatorMod implements
        PostInitializeSubscriber,
        PostRenderSubscriber,
        OnCardHoveredSubscriber {
    public static ModInfo info;
    public static String modID; //Edit your pom.xml to change this

    static {
        loadModInfo();
    }

    private static final String resourcesFolder = checkResourcesPath();
    public static final Logger logger = LogManager.getLogger(modID); //Used to output to the console.

    //This is used to prefix the IDs of various objects like cards and relics,
    //to avoid conflicts between different mods using the same name for things.
    public static String makeID(String id) {
        return modID + ":" + id;
    }

    //This will be called by ModTheSpire because of the @SpireInitializer annotation at the top of the class.
    public static void initialize() {
        new DmgCalculatorMod();
    }

    private AbstractCard hoveredCard;

    public DmgCalculatorMod() {
        CardHoverPublisher.subscribe(this);
        BaseMod.subscribe(this); //This will make BaseMod trigger all the subscribers at their appropriate times.
        logger.info(modID + " subscribed to DmgCalculationMod.");
    }

    @Override
    public void onCardHovered(AbstractCard card) {
        logger.info("Card " + card + " is hovering");
        this.hoveredCard = card;
    }

    @Override
    public void onNoCardHovering() {
        logger.info("No card hovering");
        this.hoveredCard = null;
    }

    @Override
    public void receivePostRender(SpriteBatch spriteBatch) {
        CalculateIncomingDamageRenderer.render(spriteBatch);
        CalculateOutgoingDamageRenderer.render(spriteBatch, hoveredCard);
    }

    @Override
    public void receivePostInitialize() {
        //This loads the image used as an icon in the in-game mods menu.
        Texture badgeTexture = TextureLoader.getTexture(imagePath("dmg-calculator-badge.png"));
        //Set up the mod information displayed in the in-game mods menu.
        //The information used is taken from your pom.xml file.

        //If you want to set up a config panel, that will be done here.
        //You can find information about this on the BaseMod wiki page "Mod Config and Panel".
        BaseMod.registerModBadge(badgeTexture, info.Name, GeneralUtils.arrToString(info.Authors), info.Description, null);
    }

    public static String imagePath(String file) {
        return resourcesFolder + "/images/" + file;
    }

    /**
     * Checks the expected resources path based on the package name.
     */
    private static String checkResourcesPath() {
        String name = DmgCalculatorMod.class.getName(); //getPackage can be iffy with patching, so class name is used instead.
        int separator = name.indexOf('.');
        if (separator > 0)
            name = name.substring(0, separator);

        FileHandle resources = new LwjglFileHandle(name, Files.FileType.Internal);

        if (!resources.exists()) {
            throw new RuntimeException("\n\tFailed to find resources folder; expected it to be at  \"resources/" + name + "\"." +
                    " Either make sure the folder under resources has the same name as your mod's package, or change the line\n" +
                    "\t\"private static final String resourcesFolder = checkResourcesPath();\"\n" +
                    "\tat the top of the " + DmgCalculatorMod.class.getSimpleName() + " java file.");
        }
        if (!resources.child("images").exists()) {
            throw new RuntimeException("\n\tFailed to find the 'images' folder in the mod's 'resources/" + name + "' folder; Make sure the " +
                    "images folder is in the correct location.");
        }

        return name;
    }

    /**
     * This determines the mod's ID based on information stored by ModTheSpire.
     */
    private static void loadModInfo() {
        Optional<ModInfo> infos = Arrays.stream(Loader.MODINFOS).filter((modInfo) -> {
            AnnotationDB annotationDB = Patcher.annotationDBMap.get(modInfo.jarURL);
            if (annotationDB == null)
                return false;
            Set<String> initializers = annotationDB.getAnnotationIndex().getOrDefault(SpireInitializer.class.getName(), Collections.emptySet());
            return initializers.contains(DmgCalculatorMod.class.getName());
        }).findFirst();
        if (infos.isPresent()) {
            info = infos.get();
            modID = info.ID;
        } else {
            throw new RuntimeException("Failed to determine mod info/ID based on initializer.");
        }
    }
}

File: src/main/kotlin/dmgcalculator/DmgCalculatorMod.kt
package dmgcalculator

import basemod.BaseMod
import basemod.interfaces.OnPlayerTurnStartSubscriber
import basemod.interfaces.PostInitializeSubscriber
import basemod.interfaces.PostRenderSubscriber
import com.badlogic.gdx.Files
import com.badlogic.gdx.backends.lwjgl.LwjglFileHandle
import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.evacipated.cardcrawl.modthespire.Loader
import com.evacipated.cardcrawl.modthespire.ModInfo
import com.evacipated.cardcrawl.modthespire.Patcher
import com.evacipated.cardcrawl.modthespire.lib.SpireInitializer
import dmgcalculator.interfaces.PlayerEndTurnSubscriber
import dmgcalculator.publisher.PlayerEndTurnPublisher
import dmgcalculator.renderer.MonsterRenderer
import dmgcalculator.renderer.PlayerRenderer
import dmgcalculator.util.TextureLoader.getTexture
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.*

@SpireInitializer
class DmgCalculatorMod : PostInitializeSubscriber,
    PostRenderSubscriber,
    OnPlayerTurnStartSubscriber,
    PlayerEndTurnSubscriber {

    private var isPlayerTurn: Boolean = false

    init {
        PlayerEndTurnPublisher.subscribe(this)
        BaseMod.subscribe(this) //This will make BaseMod trigger all the subscribers at their appropriate times.
        logger.info("$modID subscribed to DmgCalculationMod.")
    }

    override fun receivePostRender(spriteBatch: SpriteBatch?) {
        spriteBatch?.let {
            PlayerRenderer.render(it, isPlayerTurn)
            MonsterRenderer.render(it)
        }
    }

    override fun receiveOnPlayerTurnStart() {
        isPlayerTurn = true
    }

    override fun onPlayerEndTurn() {
        isPlayerTurn = false
    }

    override fun receivePostInitialize() {
        //This loads the image used as an icon in the in-game mods menu.
        val badgeTexture = getTexture(imagePath("dmg-calculator-badge.png"))

        //Set up the mod information displayed in the in-game mods menu.
        //The information used is taken from your pom.xml file.

        //If you want to set up a config panel, that will be done here.
        //You can find information about this on the BaseMod wiki page "Mod Config and Panel".
        BaseMod.registerModBadge(
            badgeTexture,
            info!!.Name,
            info!!.Authors?.joinToString(", "),
            info!!.Description,
            null
        )
    }

    companion object {

        private val resourcesFolder: String = checkResourcesPath()

        var info: ModInfo? = null
        var modID: String? = null //Edit your pom.xml to change this
        val logger: Logger = LogManager.getLogger(modID) //Used to output to the console.

        init {
            loadModInfo()
        }

        //This will be called by ModTheSpire because of the @SpireInitializer annotation at the top of the class.
        @JvmStatic
        fun initialize() {
            DmgCalculatorMod()
        }

        fun imagePath(file: String?): String {
            return "$resourcesFolder/images/$file"
        }

        /**
         * Checks the expected resources path based on the package name.
         */
        private fun checkResourcesPath(): String {
            var name =
                DmgCalculatorMod::class.java.getName() //getPackage can be iffy with patching, so class name is used instead.
            val separator = name.indexOf('.')
            if (separator > 0) name = name.substring(0, separator)

            val resources: FileHandle = LwjglFileHandle(name, Files.FileType.Internal)

            if (!resources.exists()) {
                throw RuntimeException(
                    "\n\tFailed to find resources folder; expected it to be at  \"resources/" + name + "\"." +
                            " Either make sure the folder under resources has the same name as your mod's package, or change the line\n" +
                            "\t\"private static final String resourcesFolder = checkResourcesPath();\"\n" +
                            "\tat the top of the " + DmgCalculatorMod::class.java.getSimpleName() + " java file."
                )
            }
            if (!resources.child("images").exists()) {
                throw RuntimeException(
                    "\n\tFailed to find the 'images' folder in the mod's 'resources/" + name + "' folder; Make sure the " +
                            "images folder is in the correct location."
                )
            }

            return name
        }

        /**
         * This determines the mod's ID based on information stored by ModTheSpire.
         */
        private fun loadModInfo() {
            val infos = Arrays.stream<ModInfo?>(Loader.MODINFOS).filter { modInfo: ModInfo ->
                val annotationDB = Patcher.annotationDBMap[modInfo.jarURL] ?: return@filter false
                val initializers = annotationDB.getAnnotationIndex()
                    .getOrDefault(SpireInitializer::class.java.getName(), mutableSetOf<String?>())
                initializers.contains(DmgCalculatorMod::class.java.getName())
            }.findFirst()
            if (infos.isPresent) {
                info = infos.get()
                modID = info!!.ID
            } else {
                throw RuntimeException("Failed to determine mod info/ID based on initializer.")
            }
        }
    }
}
